File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
from collections import deque

def solution(queue1, queue2):
    answer = 0
    q_sum = sum(queue1) + sum(queue2)
    if not q_sum % 2 == 0:
        return -1
    goal = (sum(queue1) + sum(queue2)) // 2
    
    max_change = (len(queue1) + len(queue2)) * 2
    current_sum = sum(queue1)
    
    queue1 = deque(queue1)
    queue2 = deque(queue2)
    
    # while 문안에서 계속 sum으로 구하기 보다 +-로 갱신
    while current_sum != goal:
        
        if answer > max_change:
            return -1
        
        if current_sum > goal:
            pop_num = queue1.popleft()
            queue2.append(pop_num)
            current_sum -= pop_num
            answer += 1
        elif current_sum < goal:
            pop_num = queue2.popleft()
            queue1.append(pop_num)
            current_sum += pop_num
            answer += 1
    
    return answer

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();
        int maxChange = 0;
        
        long sum1 = 0;
        for (int num : queue1) {
            sum1 += num;
            q1.offer(num);
            maxChange++;
        }
        long sum2 = 0;
        for (int num: queue2) {
            sum2 += num;
            q2.offer(num);
            maxChange++;
        }
        long sum = sum1 + sum2;
        if (sum % 2 != 0)
            return -1;
        sum /= 2;
        maxChange *= 2;
        
        int answer = 0;
        while (answer < maxChange) {
            if (sum1 == sum) {
                return answer;
            }
            else if (sum1 > sum) {
                int num = q1.poll();
                q2.offer(num);
                sum1 -= num;
            }
            else {
                int num = q2.poll();
                q1.offer(num);
                sum1 += num;
            }
            answer++;
        }
        
        return -1;
    }
}